### PR TITLE
Handle deconstruct pattern in ref safety analysis

### PR DESCRIPTION
**Pull request by jjonescz**
Fixes https://github.com/dotnet/roslyn/issues/75484.

### CODE REVIEW

**Comment by jjonescz** [comment]
@dotnet/roslyn-compiler for reviews, thanks

**Comment by jaredpar** [comment]
@jcouv, @cston PTAL

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs:6200]: nit: Consider also testing a nested deconstruction scenario: `((var x, var y), var z)`. I expect it should just fall out by composition

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 1)

**Comment by jcouv** [comment]
@cston for second review. Thanks

File: src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs:624]: I didn't understand why we should pass 'isDiscardExpression: true' here.

Comment (jjonescz) [reply src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs:624]: Because this placeholder corresponds to an `out _` discard parameter in the lowered `Deconstruct` call.

Comment (jjonescz) [reply src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs:624]: Originally this code path was written to handle discards in the pattern which correspond to `out _` discard arguments of the equivalent `Deconstruct` call.

But now I realize there can be also nested subpatterns and we need to be smarter about extracting all variable accesses from that.

**Comment by jjonescz** [comment]
@jcouv @RikkiGibson for another look, thanks

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs:6109]: It feels like we should also have tests where the Deconstruct receiver is something we can return a reference to. For example, if it is a by-ref parameter.

Comment (jjonescz) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs:6109]: Thanks, that actually doesn't work as expected.

**Comment by jjonescz** [comment]
@dotnet/roslyn-compiler for reviews, thanks

**Comment by jjonescz** [comment]
@jcouv @cston for reviews, thanks

File: src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs:603]: >{ HasUnscopedRefAttribute: true } [](http://example.com/codeflow?start=42&length=33)

nit: I'm wondering whether this could also be written in terms of `EffectiveScope`, for regularity with second case. Would something like `{ IsExtensionMethod: false, ThisParameter: { EffectiveScope: ScopedKind.None } }` work?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs:5986]: nit: it's a bit difficult to check that every scenario with `Deconstruct` invocation has an equivalent deconstruction-assignment scenario. Consider merging those pairs. 
For example, in `Deconstruction_UnscopedRef` this would mean adding:
```
                    R M2()
                    {
                        (var x, _) = new S();
                        return x;
                    }
```

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 4)

File: src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs:610]: Looking through the spec some more, makes me wonder: what if the `Deconstruct` method was compiled with an older compiler? Then the receiver, which is treated as a `ref` parameter, would be considered unscoped by default.

Here's the relevant [section](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-11.0/low-level-struct-improvements.md#refsafetyrulesattribute):
> To reduce the chance of breaking changes when recompiling with C#11, we will update the C#11 compiler to use the ref safe context rules for method invocation that match the rules that were used to analyze the method declaration. Essentially, when analyzing a call to a method compiled with an older compiler, the C#11 compiler will use C#7.2 ref safe context rules.

Comment (jjonescz) [reply src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs:610]: > Let me try to summarize my understanding. Does that match your understanding?

Yes, your summary matches my understanding.

> what if the `Deconstruct` method was compiled with an older compiler? Then the receiver, which is treated as a `ref` parameter, would be considered unscoped by default.

Good point. I think if I rewrite the condition in terms of EffectiveScope as you suggested in the other comment, and we take https://github.com/dotnet/roslyn/pull/75930, this scenario will work (because that PR makes EffectiveScope consider old/updated rules).

File: src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs
Comment (jjonescz) [reply src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs:603]: `ThisParameter` property looks like it could throw, so added a local function instead of the pattern.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs:5986]: Thanks for the change. It's much easier to follow now :-)

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 6) Let's wait for the [PR](https://github.com/dotnet/roslyn/pull/75930) related to EffectiveScope to go in and add a test for the intersection scenario that'll fix.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs:6179]:
```suggestion
                        return x;
```

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs:6410]: It wasn't clear to me why we are also testing this here, it seems the same as the `M1` case.

**Review by RikkiGibson — COMMENTED** [review]
LGTM, just want to wait on answers to questions and upstream effects from other PR

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs
Comment (jjonescz) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs:6199]: Yes, `M3` is lowered to the following which would be a ref safety error as well at the Deconstruct call:

```cs
R M3(ref S s)
{
    S s2 = s;
    R x;
    int y;
    s2.Deconstruct(out x, out y);
    if (true)
    {
        return x;
    }
    return default(R);
}
```

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs
Comment (jjonescz) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs:6410]: It's not the same - `M1` returns `x` whereas `M2` returns `x2`

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs:610]: Using `EffectiveScope` seems a good solution to abstract away from those differences.
Were you able to add a test for this old rules vs. new rules scenario?

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 7)

File: src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs
Comment (jjonescz) [reply src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs:610]: I'm waiting for the other PR to be merged before adding a test.

Comment (jjonescz) [reply src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs:610]: > what if the `Deconstruct` method was compiled with an older compiler? Then the receiver, which is treated as a `ref` parameter, would be considered unscoped by default.

Actually this does not seem to be true. It's hard to tell from the spec how it should be because the old rules have no concept of scoped/unscoped. Anyway, I added tests that verify the behavior is consistent between explicit calls and patterns, so it behaves as expected wrt this PR, I think.

**Comment by jjonescz** [comment]
@jcouv for another look, thanks

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 9)

**Pull request merged**
